File: NostrVault/app/src/main/java/com/nostrvault/data/local/CredentialStore.kt
package com.nostrvault.data.local

import android.content.Context
import android.content.SharedPreferences
import android.util.Log
import androidx.security.crypto.EncryptedSharedPreferences
import androidx.security.crypto.MasterKey

/**
 * Encrypted credential storage -- port of CredentialStore.swift.
 *
 * iOS uses Keychain Services; Android uses EncryptedSharedPreferences
 * backed by AndroidKeyStore (AES-256-GCM).
 *
 * All methods are thread-safe (SharedPreferences handles synchronization).
 */
object CredentialStore {

    private const val TAG = "CredentialStore"
    private const val PREFS_NAME = "nostrvault_credentials"
    private const val OWNER_ACCOUNT = "owner-key-password"

    private lateinit var prefs: SharedPreferences

    fun init(context: Context) {
        if (::prefs.isInitialized) return

        try {
            val masterKey = MasterKey.Builder(context.applicationContext)
                .setKeyScheme(MasterKey.KeyScheme.AES256_GCM)
                .build()
            prefs = EncryptedSharedPreferences.create(
                context.applicationContext,
                PREFS_NAME,
                masterKey,
                EncryptedSharedPreferences.PrefKeyEncryptionScheme.AES256_SIV,
                EncryptedSharedPreferences.PrefValueEncryptionScheme.AES256_GCM,
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to create EncryptedSharedPreferences: ${e.message}")
            throw e
        }
    }

    private fun accountId(npub: String): String = "account-key-password-$npub"

    // ---- Owner (global) credential ----

    fun storeOwnerPassword(password: String): Boolean =
        store(password, OWNER_ACCOUNT)

    fun getOwnerPassword(): String? =
        get(OWNER_ACCOUNT)

    fun deleteOwnerPassword(): Boolean =
        delete(OWNER_ACCOUNT)

    fun hasOwnerPassword(): Boolean =
        getOwnerPassword() != null

    // ---- Per-account credentials ----

    fun storePassword(password: String, forNpub: String): Boolean =
        store(password, accountId(forNpub))

    fun getPassword(forNpub: String): String? =
        get(accountId(forNpub))

    fun deletePassword(forNpub: String): Boolean =
        delete(accountId(forNpub))

    // ---- nsec storage (hex secret key keyed by hex pubkey) ----

    private fun nsecKey(hexPubkey: String): String = "nsec-hex-$hexPubkey"

    fun saveNsec(hexSeckey: String, hexPubkey: String): Boolean =
        store(hexSeckey, nsecKey(hexPubkey))

    fun getNsec(hexPubkey: String): String? =
        get(nsecKey(hexPubkey))

    // ---- Keychain password (NIP-49 passphrase, keyed by npub) ----

    private fun keychainKey(npub: String): String = "keychain-$npub"

    fun storeKeychainPassword(password: String, npub: String): Boolean =
        store(password, keychainKey(npub))

    fun getKeychainPassword(npub: String): String? =
        get(keychainKey(npub))

    // ---- Whitelisted account hex keys ----

    private fun credHexKey(npub: String): String = "cred-hex-$npub"

    fun storeCredentialHexKey(hexKey: String, npub: String): Boolean =
        store(hexKey, credHexKey(npub))

    fun getCredentialHexKey(npub: String): String? =
        get(credHexKey(npub))

    // ---- FIPS mesh network identity ----
    //
    // Deliberately not one of the account keys above: the mesh address should
    // be rotatable without touching the social identity, and vice versa. One
    // per install, not per account.

    private const val MESH_IDENTITY = "fips-mesh-nsec"

    fun storeMeshNsec(nsec: String): Boolean =
        store(nsec, MESH_IDENTITY)

    fun getMeshNsec(): String? =
        get(MESH_IDENTITY)

    fun deleteMeshNsec(): Boolean =
        delete(MESH_IDENTITY)

    // ---- Private helpers ----

    private fun store(password: String, account: String): Boolean = try {
        prefs.edit().putString(account, password).commit()
    } catch (e: Exception) {
        Log.e(TAG, "Failed to store credential for $account: ${e.message}")
        false
    }

    private fun get(account: String): String? = try {
        prefs.getString(account, null)
    } catch (e: Exception) {
        Log.e(TAG, "Failed to retrieve credential for $account: ${e.message}")
        null
    }

    private fun delete(account: String): Boolean = try {
        prefs.edit().remove(account).commit()
    } catch (e: Exception) {
        Log.e(TAG, "Failed to delete credential for $account: ${e.message}")
        false
    }
}

File: HavenApp/HavenApp/Services/CredentialStore.swift
import Foundation
import Security

/// Keychain credential storage — platform-specific implementation.
/// Android equivalent: AndroidKeyStore / EncryptedSharedPreferences
/// with the same method signatures and semantics.
///
/// All methods are static and thread-safe (Security framework serialises
/// Keychain access internally).
enum CredentialStore {

    private static let service = "com.havenapp.nip49"
    private static let ownerAccount = "owner-key-password"

    private static func accountId(forNpub npub: String) -> String {
        "account-key-password-\(npub)"
    }

    // MARK: - Owner (global) credential

    @discardableResult
    static func storeOwnerPassword(_ password: String) -> Bool {
        store(password, account: ownerAccount)
    }

    static func getOwnerPassword() -> String? {
        get(account: ownerAccount)
    }

    @discardableResult
    static func deleteOwnerPassword() -> Bool {
        delete(account: ownerAccount)
    }

    static func hasOwnerPassword() -> Bool {
        getOwnerPassword() != nil
    }

    // MARK: - Per-account credentials

    @discardableResult
    static func storePassword(_ password: String, forNpub npub: String) -> Bool {
        store(password, account: accountId(forNpub: npub))
    }

    static func getPassword(forNpub npub: String) -> String? {
        get(account: accountId(forNpub: npub))
    }

    @discardableResult
    static func deletePassword(forNpub npub: String) -> Bool {
        delete(account: accountId(forNpub: npub))
    }

    // MARK: - FIPS mesh network identity

    // Deliberately separate from any Nostr key: the mesh address should be
    // rotatable without touching the social identity. Stored here rather than
    // in HavenConfig because it is a secret, and because the file it would
    // otherwise land in is synced and backed up.
    private static let meshNsecAccount = "fips-mesh-nsec"

    @discardableResult
    static func storeMeshNsec(_ nsec: String) -> Bool {
        store(nsec, account: meshNsecAccount)
    }

    static func getMeshNsec() -> String? {
        get(account: meshNsecAccount)
    }

    @discardableResult
    static func deleteMeshNsec() -> Bool {
        delete(account: meshNsecAccount)
    }

    // MARK: - Private helpers

    private static func store(_ password: String, account: String) -> Bool {
        guard let data = password.data(using: .utf8) else { return false }

        let deleteQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        SecItemDelete(deleteQuery as CFDictionary)

        let addQuery: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecValueData as String: data,
            kSecAttrAccessible as String: kSecAttrAccessibleAfterFirstUnlock,
        ]
        return SecItemAdd(addQuery as CFDictionary, nil) == errSecSuccess
    }

    private static func get(account: String) -> String? {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
            kSecReturnData as String: true,
            kSecMatchLimit as String: kSecMatchLimitOne,
        ]
        var result: AnyObject?
        let status = SecItemCopyMatching(query as CFDictionary, &result)
        if status == errSecSuccess, let data = result as? Data {
            return String(data: data, encoding: .utf8)
        }
        return nil
    }

    private static func delete(account: String) -> Bool {
        let query: [String: Any] = [
            kSecClass as String: kSecClassGenericPassword,
            kSecAttrService as String: service,
            kSecAttrAccount as String: account,
        ]
        let status = SecItemDelete(query as CFDictionary)
        return status == errSecSuccess || status == errSecItemNotFound
    }
}
